plugins/meta/sbr/README.md

```diff
@@ -0,0 +1,131 @@
+passed in from the previous plugin (which has just set up the network device):
+
+- What is the network interface in question?
+
+- What is the IP address (or addresses) of that network interface?
+
+- What is the default gateway for that interface (if any)?
+
+It then reads all routes to the network interface in use.
+
+Here is an example of what the plugin would do. (The `ip` based commands are
+implemented in go, but easier to describe via the command line.) Suppose that
+it reads that:
+
+- The interface is `net1`.
+
+- The IP address on that interface is `192.168.1.209`.
+
+- The default gateway on that interface is `192.168.1.1`.
+
+- There is one route configured on that network interface, which is
+  `192.168.1.0/24`.
+
+Then the actions it takes are the following.
+
+- It creates a new routing table, and sets a rule to use it for the IP address in question.
+
+        ip rule add from 192.168.1.209/32 table 100
+
+- It adds a route to the default gateway for the relevant table.
+
+        ip route add default via 192.168.1.1 dev net1 table 100
+
+- It moves every existing route on the device to the new table.
+
+        ip route del 192.168.1.0/24 dev net1 src 192.168.1.209
+        ip route add 192.168.1.0/24 dev net1 src 192.168.1.209 table 100
+
+On deletion it:
+
+- deletes the rule (`ip rule del from 192.168.1.209/32 table 100`), which it
+  finds by deleting rules relating to IPs on the device which is about to be
+  deleted.
+
+- does nothing with routes (since the kernel automatically removes routes when
+  the device with which they are associated is deleted).
+
+## Future enhancements and known limitations
+
+The following are possible future enhancements.
+
+- The table number is currently selected by starting at 100, then incrementing
+  the value until an unused table number is found. It might be nice to have an
+  option to pass the table number as an input.
+
+- The logging is entirely on or entirely off - there is no log severity.
+
+- This plugin sets up Source Based Routing, as described above. In future,
+  there may be a need for a VRF plugin (that uses
+  [VRF routing](https://www.kernel.org/doc/Documentation/networking/vrf.txt)
+  instead of source based routing). If and when this happens, it is likely that
+  the logic would be virtually identical for the plugin, and so the same plugin
+  might offer either SBR or VRF routing depending on configuration.
+
+## Configuration
+
+This plugin must be used as a chained plugin. There is a single configuration
+parameter.
+
+- If `logfile` is set, then logs are written to that log file on
+  disk. Otherwise all logs are written to `stderr`.
+
+A sample configuration for this plugin acting as a chained plugin after flannel
+is the following.
+
+~~~json
+{
+  "name": "flannel-sbr",
+  "cniVersion": "0.3.0",
+  "plugins":
+    [
+        { "type": "flannel" },
+        {
+          "type": "sbr",
+          "logfile": "/var/log/sbr.log"
```

Comment:
Generally, plugins just log to stderr.

Reply:
True, but I'd like to appeal on this one. By default the plugin just logs to stderr, but can be configured to log to file. I actually found the option of setting that debug option was very useful when testing.

Hence I can remove the logging to file if you want, but it felt like a good thing to have available when I needed it. What do you think?

Reply:
@plwhite we don't have any general objection to logging elsewhere than stderr, but we believe plugins should always log to stderr. Any other logging should be additional to stderr. But we also want to figure out a "standard" way to tell plugins what level to log at and what the additional method to use, since we don't think that should be different for every plugin.

(eg, you would just want to tell Kubernetes or CRIO or Mesos to "turn on CNI debugging" and it would work for all plugins in a chain)

Reply:
@plwhite so for now, if it's OK, we'd like to remove the logging and add it back later when we have a better plan for logging in general.

Reply:
Having thought this over, I agree that every plugin logging in its own way is insane, and strictly it's way more sensible that K8s should be trapping stderr and saving it somewhere.
So fair enough, I have removed it in my latest commit.